Add quota check for FileSystem writables

youennf opened the pull request:
#### a33fd49e6aeff21a4e139088fc38ddcec176d78e
<pre>
Add quota check for FileSystem writables
<a href="https://rdar.apple.com/144381471">rdar://144381471</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287238">https://bugs.webkit.org/show_bug.cgi?id=287238</a>

Reviewed by Sihui Liu.

Add a quota check when writing data, and for truncate since truncate can increase the file size.
We also do a zero quota check for other commands to make sure the order of operation is preserved.

Covered by added test.

* LayoutTests/storage/filesystemaccess/writable-quota-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-quota.html: Added.
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.h:
(WebKit::convertToException):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.serialization.in:
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::executeCommandForWritableInternal):
(WebKit::FileSystemStorageHandle::executeCommandForWritable):
(WebKit::FileSystemStorageHandle::requestNewCapacityForSyncAccessHandle):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::executeCommandForWritable):

Canonical link: <a href="https://commits.webkit.org/291304@main">https://commits.webkit.org/291304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd8b4837b1135679a35db7c0c50c38ff8f5122bd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1534 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97384 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42907 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20402 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70815 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28274 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95402 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83683 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51133 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1311 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42237 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79321 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1261 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99410 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19451 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14378 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79826 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19701 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79556 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79082 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19632 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23628 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/928 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12452 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19435 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24607 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19122 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22582 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20862 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
```diff
@@ -346,13 +345,64 @@ std::optional<FileSystemStorageError> FileSystemStorageHandle::executeCommandFor
-std::optional<FileSystemStorageError> FileSystemStorageHandle::executeCommandForWritable(WebCore::FileSystemWritableFileStreamIdentifier streamIdentifier, WebCore::FileSystemWriteCommandType type, std::optional<uint64_t> position, std::optional<uint64_t> size, std::span<const uint8_t> dataBytes, bool hasDataError)
+void FileSystemStorageHandle::executeCommandForWritable(WebCore::FileSystemWritableFileStreamIdentifier streamIdentifier, WebCore::FileSystemWriteCommandType type, std::optional<uint64_t> position, std::optional<uint64_t> size, std::span<const uint8_t> dataBytes, bool hasDataError, CompletionHandler<void(std::optional<FileSystemStorageError>)>&& completionHandler)
 {
+    if (type == WebCore::FileSystemWriteCommandType::Write && !hasDataError) {
```
I remember file size can be changed for [truncate()](https://fs.spec.whatwg.org/#api-filesystemwritablefilestream-truncate), so I think we should add quota check for that too.

szewai commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
```diff
@@ -346,13 +345,74 @@ std::optional<FileSystemStorageError> FileSystemStorageHandle::executeCommandFor
+            finalSize = *size;
+
+        if (finalSize > *fileSize) {
+            RefPtr manager = m_manager.get();
```
We should probably null check `m_manager` before doing the size calculation

szewai commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
```diff
@@ -346,13 +345,74 @@ std::optional<FileSystemStorageError> FileSystemStorageHandle::executeCommandFor
+            completionHandler(FileSystemStorageError::Unknown);
+            return;
+        }
+        uint64_t finalSize;
```
Nit: `uint64_t finalSize = *fileSize;` and we don't need to do `else` statement below.

szewai commented on Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
```diff
@@ -346,13 +345,74 @@ std::optional<FileSystemStorageError> FileSystemStorageHandle::executeCommandFor
+                completionHandler(error);
+            });
+            return;
+        }
```
oh do we need to make sure write request that does not need additional space is also sent to quota manager (quota check queue) with size 0 to keep ordering of write operations? 
We might want to add a test for that.

youennf committed:
Add quota check for FileSystem writables
rdar://144381471
https://bugs.webkit.org/show_bug.cgi?id=287238

Reviewed by Sihui Liu.

Add a quota check when writing data, and for truncate since truncate can increase the file size.
We also do a zero quota check for other commands to make sure the order of operation is preserved.

Covered by added test.

* LayoutTests/storage/filesystemaccess/writable-quota-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-quota.html: Added.
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.h:
(WebKit::convertToException):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.serialization.in:
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::executeCommandForWritableInternal):
(WebKit::FileSystemStorageHandle::executeCommandForWritable):
(WebKit::FileSystemStorageHandle::requestNewCapacityForSyncAccessHandle):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::executeCommandForWritable):

Canonical link: https://commits.webkit.org/291304@main
